src/cli/build.rs

```diff
@@ -0,0 +1,84 @@
+use console::Style;
+use std::{
+    env,
+    path::{Path, PathBuf},
+    process::ExitCode,
+};
+use tokio::{
+    fs::{self, OpenOptions},
+    io::AsyncWriteExt,
+};
+
+use anyhow::Result;
+use mlua::Compiler as LuaCompiler;
+
+/**
+    Compiles and embeds the bytecode of a requested lua file to form a standalone binary,
+    then writes it to an output file, with the required permissions.
+*/
+#[allow(clippy::similar_names)]
+pub async fn build_standalone<T: AsRef<Path> + Into<PathBuf>>(
+    script_path: String,
+    output_path: T,
+    code: impl AsRef<[u8]>,
+) -> Result<ExitCode> {
+    let log_output_path = output_path.as_ref().display();
+
+    let prefix_style = Style::new().green().bold();
+    let compile_prefix = prefix_style.apply_to("Compile");
+    let bytecode_prefix = prefix_style.apply_to("Bytecode");
+    let write_prefix = prefix_style.apply_to("Write");
+    let compiled_prefix = prefix_style.apply_to("Compiled");
+
+    println!("{compile_prefix} {script_path}");
+
+    // First, we read the contents of the lune interpreter as our starting point
+    let mut patched_bin = fs::read(env::current_exe()?).await?;
+    let base_bin_offset = u64::try_from(patched_bin.len())?;
+
+    // The signature which separates indicates the presence of bytecode to execute
+    // If a binary contains this signature, that must mean it is a standalone binary
+    let signature: Vec<u8> = vec![0x4f, 0x3e, 0xf8, 0x41, 0xc3, 0x3a, 0x52, 0x16];
```

Comment:
No, we cannot, if we stored the signature as a constant, we'd need it to be a value which is sized, meaning an array slice. Array slices are written to the stack of the binary, meaning all binaries would include this signature, regardless of being standalone or not.

Reply:
> Array slices are written to the stack of the binary

I'm not quite sure what you mean by this, but if you are saying that the literal bytes for the array slice would be present in a contiguous manner in the resulting binary, potentially leading to misidentifying non-standalone binaries, they are not (https://godbolt.org/z/1v3Gdfo9r) since each byte will get interleaved with normal assembly instructions.
We are also not checking the entire binary for this signature, just the end. The collision probability for 8 bytes, assuming generated assembly is random bytes, is something like 256^8, so we'd be fine with like 6+ bytes (one in a quadrillion chance)

Reply:
Oh right! Array slices *are indeed* written to the stack of the binary, literally, based on my experiences with the hex of the binary, but I totally missed the fact that we're only checking the last few bytes (which I initially wasn't, when I started writing this code).

Furthermore, the fact about the array slices being interleaved by assembly instructions is irrelevant, since the assembly instructions are just representations of the machine code of the binary, not the literal instructions in the binary.

Reply:
I was not able to find any traces of it being in the binary using the code I linked, but, I did find this: 
![CleanShot 2024-01-04 at 17 17 44@2x](https://github.com/lune-org/lune/assets/9497781/cd91bc83-4ae1-447e-a5e4-1c00299cca74)
This is what tells me that it's not just dumping the slice as-is into the binary - I'm finding the bytes, yeah, but they're aligned to word/instruction size - maybe what you're seeing is debug symbols? Or it being w...i...d...e... for me might be a platform/arch specific thing?

Reply:
Tried a release build and can still replicate. 

![image](https://github.com/lune-org/lune/assets/74418041/d4f8305e-a3f4-4572-aa01-3c2dc0e8515f)

Reply:
Just gave your example code a read :P

Right, the reason the bytes in your case are padded is due to the fact that you're getting the bytes of a UTF-8 string, which are always padded to uniform alignment. So, this proves the fact that arrays/array slices will be written to the stack of the binary regardless.

Reply:
It's a byte string literal (has a `b` prefix) so no utf8 there https://doc.rust-lang.org/reference/tokens.html#byte-string-literals

Was also a bit confused about why the assembly instructions were exactly 10 movs with the actual bytes, if the type is not sized this should not be possible 🤔 turns out the `vec!` macro really just expands into a static slice, gets boxed for the allocation, and coerced into a vec from that: https://github.com/rust-lang/rust/blob/090d5eac722000906cc00d991f2bf052b0e388c3/library/alloc/src/macros.rs#L42-L57

Reply:
Oh, that's quite interesting, guess I was quite off then.

Reply:
If you *really* want to store it as a constant, we could have the signature we check against be scrambled a bunch with bitwise ops, and then unscramble it at runtime before checking for the signature in the binary.

Reply:
I don't think any scrambling is necessary - magic/signature bytes will only ever be at the end, and as far as I'm aware constants/debug symbols are never stored at the end of any executables. We shouldn't sacrifice readability/debuggability unless there's a tangible benefit to collision resistance for the signature.

For reference, the standalone compilation that Deno has had for some years works exactly like this, it writes a defined constant plainly to the end of the compiled file, and has never had any issues with that:
https://github.com/denoland/deno/blob/main/cli/standalone/binary.rs#L49 https://github.com/denoland/deno/blob/main/cli/standalone/binary.rs#L236-L237
https://github.com/denoland/deno/blob/main/cli/standalone/binary.rs#L330

Reply:
Will update to make it a constant then.

Reply:
See https://github.com/lune-org/lune/pull/140/commits/3c2464d3ce86ab31d58db996855be35165af304c.